C/unreachable.h

```diff
@@ -0,0 +1,17 @@
+/* This module defines an UNREACHABLE macro that calls '__builtin_unreachable' if it exists.
+ * Placing UNREACHABLE after a branch can allow an optimizer to eliminate the branch.
+ */
+#ifndef UNREACHABLE
```

Comment:
I wonder if we should require a `#define` to make this actually map to `__builtin_unreachable`, and have it map to a "output to stderr then abort" otherwise. `__builtin_unreachable` is scary because it's actually UB to call, so if it's ever used mistakenly then crazy things will happen.

Reply:
FWIW, some inspection of the output of GCC 8 (in a variant of this code) shows that using a `__builtin_unreachable` does successfully optimise code by removing evaluation on conditions that lead to unreachable code.

It's probably fair to perform a `DEBUG` check here to write a message to stderr, but otherwise as it stands in an `NDEBUG` build will already do crazy undefined things if invariants get violated.

Reply:
Filed https://github.com/ElementsProject/simplicity/issues/12 to rework this and the assertions to keep safety during production builds.